Documentation/eliminating-pre-builts.md

```diff
@@ -27,42 +50,159 @@ To protect against this and catch any new pre-builts as soon as possible, Arcade
+  </Usages>
+</UsageData>
+```
 
-### Pre-built through a transitive dependency
+The `project.assets.json` file shows the following:
 
-During a project dependency updated, a new pre-built might be introduced by a new or updated transitive dependency. While the Arcade tooling will highlight the name and version of the pre-built in the build exception as well as the project that restored the dependency in question, it will not point out where exactly in the dependency tree it is. This means that it's up to the developer to identify what type of dependency they are dealing with (direct or transitive) and choose the correct way of handling the issue. In case of a transitive dependency, it might be hard to identify the relationship that is bringing in the pre-built into the project, especially if the developer has limited knowledge of the project or code-base in general.
+```json
+{ 
+  "targets": {
+    "net8.0": {
+      "System.Text.Json/8.0.0": {
+        "type": "package",
+        "dependencies": {
+        }
+      }
+    }
+  }
+}
+```
 
-Arcade source-build infrastructure helps accomplish this this by pointing out the `project.assets.json` file that is referencing the pre-built in the `./artifacts/source-build/self/prebuild-report/prebuild-usage.xml` report file. A `project.assets.json` file is a NuGet restore process artifact that contains a resolved dependency tree for a specific project. Every package that was restored by a given project is mentioned there with links between the dependencies, allowing the reader to identify transitive dependencies and the direct dependencies referencing them.
+You find a package reference to `System.Text.Json.8.0.0`, like the one above, in one of your .csproj files:
 
-Example of a pre-built caused by a transitive dependency and corresponding entries in files mentioned above:
+```xml
+<ItemGroup>
+    <PackageReference Include="System.Text.Json" Version="8.0.0" />
+</ItemGroup>
+```
 
-Exception identifying the pre-built:
+In this example, `System.Text.Json` is a direct dependency causing a pre-built.
+
+At this point, determine how to handle the dependency by following the [adding dependencies documentation](https://github.com/dotnet/source-build/blob/main/Documentation/sourcebuild-in-repos/new-dependencies.md#adding-dependencies). Based on what you determine, you will likely have to do one or more of the following:
+
+- **Update version -** Update the version of System.Text.Json to match the version of the repository that the package is flowing in from.
+- **Use Version.Details and Version.Props -**
+  - Add an entry to `./eng/Version.Details.xml` for System.Text.Json
+  - Define/use `<SystemTextJsonVersion>8.0.0</SystemTextJsonVersion>` in `./eng/Versions.Props`
+- **Source-build-reference-packages -** Add `System.Text.Json.8.0.0` to the [source-build-reference-packages](https://github.com/dotnet/source-build-reference-packages/) repo.
```

Comment:
Somewhere we need to state that adding a package to SBRP isn't always the right step to take to resolve a prebuilt. For example, the repo may have added a dependency on a package and included that in Version.Details and it will give them a detected prebuilt error. Then they may think it needs to be added to SBRP. But in the context of the VMR, if the dependency comes from another repo that gets built before their repo, the package in SBRP is not needed because the VMR will flow the live version. In this case we don't want it added to SBRP and it should rather be marked as an exclusion in the baseline. I feel this is an important scenario that needs to be described.

Reply:
I'm wondering if this should be added as part of https://github.com/dotnet/source-build/blob/main/Documentation/sourcebuild-in-repos/new-dependencies.md#adding-dependencies  and then pointed to in this document as a short warning/note about the importance of building the repo in the VMR when detecting prebuilts. The reasoning behind this suggestion is that the eliminating document suggests steps to determine the causes of prebuilts, but it does not go into depth on how to resolve them. Instead, I was leaving that work up the adding-dependencies documentation.

As you pointed out in a different comment above, that documentation does need refactoring. In applying this suggestion, I think that the updates to the adding-dependencies doc can be part of this PR.

Reply:
Added here: https://github.com/dotnet/source-build/pull/4043/commits/010a2cca0ff2703d2884e1ed017123b41f216e57#diff-d4752b9dd717bdd56774b2d6f54ec8365f9cafd10917354a454278a917f0a104R93